test/unit/remote-config/remote-config.spec.ts

```diff
@@ -951,6 +954,82 @@ describe('RemoteConfig', () => {
       });
     });
 
+    describe('set', () => {
+      const INVALID_PARAMETERS: any[] = [null, '', 'abc', 1, true, []];
+      const INVALID_CONDITIONS: any[] = [null, '', 'abc', 1, true, {}];
+      let sourceTemplate = deepCopy(SERVER_REMOTE_CONFIG_RESPONSE);
+
+      it('should set template when passed', () => {
+        const template = deepCopy(SERVER_REMOTE_CONFIG_RESPONSE) as ServerTemplateData;
+        template.parameters = {
+          dog_type: {
+            defaultValue: {
+              value: 'shiba'
+            },
+            description: 'Type of dog breed',
+            valueType: 'STRING'
+          }
+        };
+        const initializedTemplate = remoteConfig.initServerTemplate();
+        initializedTemplate.set(template);
+        const parsed = initializedTemplate.toJSON();
+        const expectedVersion = deepCopy(VERSION_INFO);
+        expectedVersion.updateTime = new Date(expectedVersion.updateTime).toUTCString();
+        template.version = expectedVersion as Version;
+        expect(parsed).deep.equals(deepCopy(template));
+      });
+
+      it('should set and instantiates template when json string is passed', () => {
+        const template = deepCopy(SERVER_REMOTE_CONFIG_RESPONSE) as ServerTemplateData;
+        template.parameters = {
+          dog_type: {
+            defaultValue: {
+              value: 'shiba'
+            },
+            description: 'Type of dog breed',
+            valueType: 'STRING'
+          }
+        };
+        const templateJson = JSON.stringify(template);
+        const initializedTemplate = remoteConfig.initServerTemplate();
+        initializedTemplate.set(templateJson);
+        const parsed = initializedTemplate.toJSON();
+        const expectedVersion = deepCopy(VERSION_INFO);
+        expectedVersion.updateTime = new Date(expectedVersion.updateTime).toUTCString();
+        template.version = expectedVersion as Version;
+        expect(parsed).deep.equals(deepCopy(template));
+      });
+
+      it('should throw if template is an invalid JSON', () => {
+        const jsonString = '{invalidJson: null}';
+        const initializedTemplate = remoteConfig.initServerTemplate();
+        expect(() => initializedTemplate.set(jsonString))
+          .to.throw(/Failed to parse the JSON string: ([\D\w]*)\./);
+      });
+
+      INVALID_PARAMETERS.forEach((invalidParameter) => {
+        sourceTemplate.parameters = invalidParameter;
```

Comment:
it's unclear what this is testing. All the values in `INVALID_PARAMETERS` would produce valid JSON. For example, '{"parameters":null}' is valid.

Since we're currently also catching errors from `ServerTemplateDataImpl`, and `parameters: null` is an invalid template, I get the impression this is testing that.

I put a comment above about moving `ServerTemplateDataImpl` outside the try/catch, to differentiate JSON and template parsing errors.

Reply:
Moved `ServerTemplateDataImpl` to outside the try/catch and updated the tests here to differentiate JSON and template parsing errors